server.go

```diff
@@ -352,27 +364,84 @@ func serveFaucet(c *gin.Context) {
 		return
 	}
 
-	// Ensure the user isn't spamming the faucet
-	if user.MostRecentFaucetGrant.Add(env.FaucetRateLimit).After(time.Now()) {
-		c.JSON(http.StatusForbidden, gin.H{"error": fmt.Sprintf("you may only use the faucet once every %v hours", env.FaucetRateLimit.Hours())})
+	// No account less than a week old is allowed any FIL
+	if !user.HasAccountOlderThan(env.FaucetMinAccountAge) {
+		c.JSON(http.StatusForbidden, gin.H{"error": ErrUserTooNew.Error()})
 		return
 	}
 
-	balance, err := lotusCheckBalance(ctx, targetAddr)
-	if err != nil {
+	// returns an ID address if its a miner address, otherwise an empty address
+	minerAddr, err := lotusGetMinerAddr(ctx, targetAddr)
+	if err != nil && errors.Cause(err) != ErrNotMiner {
 		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
 		return
 	}
 
-	owed := types.BigSub(types.BigInt(env.MaxAllowanceFIL), types.BigInt(balance))
-	dif := types.BigCmp(owed, types.NewInt(0))
+	isMiner := !minerAddr.Empty()
 
-	if dif == 0 || dif == -1 {
-		c.JSON(http.StatusForbidden, gin.H{"error": "Address already has Filecoin."})
+	// ensure the non-miner/new miner hasn't already gotten their non-miner faucet tx
+	if !isMiner && user.ReceivedNonMinerFaucetGrant {
+		c.JSON(http.StatusForbidden, gin.H{"error": "non-miners can only use the faucet once"})
 		return
 	}
 
-	cid, err := lotusSendFIL(ctx, faucetAddr, targetAddr, types.FIL(owed))
+	// determine the grant size:
+	//    for non-miners: the base rate
+	//    for miners who are requesting for the first time: the base rate
+	//    for miners who have already requested at least once:
+	//    - Get the prevPower associated with the miner at last renewal
+	//    - Get the currentPower of the miner
+	//    - subtract prevPower - currentPower to get the powerDiff
+	//    - grant size = powerDiff (in GiB) / 2
+	var owed types.FIL
+	if isMiner {
+		if user.HasRequestedFromFaucetAsMiner() {
+			if user.MostRecentMinerFaucetGrant.Add(env.FaucetRateLimit).After(time.Now()) {
+				c.JSON(http.StatusForbidden, gin.H{"error": fmt.Sprintf("you may only use the faucet once every %v hours", env.FaucetRateLimit.Hours())})
+				return
+			}
+
+			targetAddr = minerAddr
+
+			receipt, err := api.StateSearchMsg(ctx, user.MostRecentMinerFaucetGrantCid)
+			if err != nil {
+				c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+				return
+			}
+
+			prevPower, err := lotusGetMinerPower(ctx, minerAddr, receipt.TipSet)
+			if err != nil {
+				c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+				return
+			}
+
+			currentPower, err := lotusGetMinerPower(ctx, minerAddr, types.EmptyTSK)
```

Comment:
Does the `tipsetKey` here _need_ to be ChainHead like you had before? I didn't think so, but let me know...

Reply:
Not getting any errors doing it this way, but let's keep this in mind if we encounter bugs